src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -369,18 +351,27 @@ protected override bool ConvertInsufficientExecutionStackExceptionToCancelledByS
 
         protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
         {
-            if (_returnTypesOpt != null)
+            if (!_returnTypesOpt.IsDefault)
             {
-                _returnTypesOpt.Clear();
+                _returnTypesOpt = _returnTypesOpt.Clear();
             }
             this.Diagnostics.Clear();
             ParameterSymbol methodThisParameter = MethodThisParameter;
             this.regionPlace = RegionPlace.Before;
-            EnterParameters(); // assign parameters
+            var parametersBuilder = PooledDictionary<Symbol, TypeWithAnnotations>.GetInstance();
+            foreach (var (key, value) in _variableTypes)
+            {
+                parametersBuilder.Add(key, value);
+            }
+            EnterParameters(parametersBuilder); // assign parameters
             if (!(methodThisParameter is null))
             {
-                EnterParameter(methodThisParameter, methodThisParameter.TypeWithAnnotations);
+                EnterParameter(methodThisParameter, methodThisParameter.TypeWithAnnotations, parametersBuilder);
             }
+            _variableTypes = parametersBuilder.ToImmutableDictionaryAndFree();
```

Comment:
>_variableTypes = parametersBuilder.ToImmutableDictionaryAndFree(); [](start = 12, length = 66)

Is the motivation for using the builder to add/update several entries in bulk? Is it worth the complexity?